### PR TITLE
(HC-41) Port Merge-related classes and ConfigValue::merge* methods

### PR DESCRIPTION
**Pull request by whopper**
The first of these commits ports skeletons of the configDelayedMergeObject
and configDelayedMerge classes, as well as the mergeableValue
interface. These are all needed for the work in the second commit, which ports each of the config  merge-related methods of ConfigValue, which are needed by the ConfigConcatenation class. They include:

```
ConfigValue::mergedWithTheUnmergeable,
ConfigValue::MergedWithObject,
ConfigValue::MergedWithNonObject,
ConfigValue::ConstructDelayedMerge,
ConfigValue::withFallback
```

Also ported are various other things necessary for ConfigConcatenation, including:

```
SimpleConfigOrigin::mergeOrigins
SimpleConfigOrigin::mergeTwo
ConfigObject::mergeOrigins
ConfigImplUtil class
```

### CODE REVIEW

**Comment by whopper** [comment]
Note: https://github.com/puppetlabs/cpp-hocon/pull/24 needs to be merged first and this rebased.

File: lib/inc/internal/config_impl_util.hpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+namespace hocon {
+
+    class config_impl_util {
+    public:
+        static bool equals_handling_null(std::string const& a, std::string const& b);
```
Comment (whopper) [inline lib/inc/internal/config_impl_util.hpp]: Question here: the original Java implementation of these methods is generic:

```
    static boolean equalsHandlingNull(Object a, Object b) {
        if (a == null && b != null)
            return false;
        else if (a != null && b == null)
            return false;
        else if (a == b) // catches null == null plus optimizes identity case
            return true;
        else
            return a.equals(b);
    }
```

From what I've seen so far, it looks like we're only going to do this with strings and vectors, so I opted to just overload it for those types. Is this okay, or do we need to take a more generic approach?

Comment (MikaelSmith) [inline lib/inc/internal/config_impl_util.hpp]: Since this is only used for strings and vectors, I'm not sure it's needed. They'll never be null because we're using references, so we can just do `a == b` and call it a day.

File: lib/inc/hocon/config_value.hpp [hunk 2]
```diff
@@ -133,6 +136,8 @@ namespace hocon {
         virtual resolve_status get_resolve_status() const;
 
         friend resolve_status resolve_status_from_values(std::vector<shared_value> const& v);
+        // this is only overridden to change the return type
+        shared_value with_fallback(std::shared_ptr<const config_mergeable> mergeable);
```
Comment (MikaelSmith) [inline lib/inc/hocon/config_value.hpp]: config_value doesn't inherit from anything, so what's it overriding?

File: lib/inc/hocon/config_object.hpp [hunk 3]
```diff
@@ -27,6 +28,9 @@ namespace hocon {
         virtual shared_object with_value(path raw_path, shared_value value) const = 0;
         virtual shared_object with_value(std::string key, shared_value value) const = 0;
 
+        shared_value with_fallback(std::shared_ptr<const config_mergeable> mergeable);
+        static shared_origin merge_origins(std::vector<shared_value> stack);
```
Comment (MikaelSmith) [inline lib/inc/hocon/config_object.hpp]: mergeOrigins is part of the internal implementation, so it should go in the `protected` block.

File: lib/inc/hocon/config_mergeable.hpp [hunk 4]
```diff
@@ -6,6 +6,7 @@
 namespace hocon {
 
     class LIBCPP_HOCON_EXPORT config_mergeable {
+        friend class config_value;
```
Comment (MikaelSmith) [inline lib/inc/hocon/config_mergeable.hpp]: Why was this needed?

Comment (whopper) [inline lib/inc/hocon/config_mergeable.hpp]: Because of one place in config_value where we do:

```
auto other = dynamic_pointer_cast<const config_mergeable>(mergeable)->to_fallback_value();
```

and if config_value isn't a friend class, we get:

```
/Users/whopper/Coding/cpp-hocon/lib/src/values/config_value.cc:124:83: error: 'to_fallback_value' is a protected member of 'hocon::config_mergeable'
            auto other = dynamic_pointer_cast<const config_mergeable>(mergeable)->to_fallback_value();
                                                                                  ^
/Users/whopper/Coding/cpp-hocon/lib/inc/hocon/config_mergeable.hpp:68:30: note: can only access this member on an object of type 'hocon::config_value'
        virtual shared_value to_fallback_value() const = 0;
```

Comment (whopper) [inline lib/inc/hocon/config_mergeable.hpp]: @MikaelSmith that cast and/or statement may also just be wrong :P

Comment (MikaelSmith) [inline lib/inc/hocon/config_mergeable.hpp]: Ok, that maybe makes sense. I'm going to play with this a bit, we shouldn't need to `friend class` when calling a method of a class the method inherits from.

Comment (MikaelSmith) [inline lib/inc/hocon/config_mergeable.hpp]: http://stackoverflow.com/a/4672888 answers the friend class requirement. We can revisit later if we'd like to just make some methods public instead of using friend everywhere.

File: lib/inc/internal/config_delayed_merge.hpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+#pragma once
+
+namespace hocon {
+
+    class config_delayed_merge : public config_value, public unmergeable {
+    public:
+        config_delayed_merge(shared_origin origin, std::vector<shared_value> stack);
+
+        config_value_type value_type() const override;
+        std::vector<shared_value> unmerged_values() const override;
```
Comment (MikaelSmith) [inline lib/inc/internal/config_delayed_merge.hpp]: I wonder if this (and thus the unmergeable interface) should return a `const&`. Not clear yet.

File: lib/inc/internal/simple_config_origin.hpp [hunk 6]
```diff
@@ -31,10 +32,28 @@ namespace hocon {
 
         std::shared_ptr<const simple_config_origin> append_comments(std::vector<std::string> comments) const;
 
+        static shared_origin merge_origins(shared_origin a, shared_origin b);
+        static shared_origin merge_origins(std::vector<shared_value> stack);
```
Comment (MikaelSmith) [inline lib/inc/internal/simple_config_origin.hpp]: These vectors should be passed by `const&`.

Comment (whopper) [inline lib/inc/internal/simple_config_origin.hpp]: Updated.

File: lib/inc/internal/simple_config_origin.hpp [hunk 7]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <hocon/config_origin.hpp>
+#include <hocon/config_value.hpp>
```
Comment (MikaelSmith) [inline lib/inc/internal/simple_config_origin.hpp]: Why is this header file needed? It wasn't before, and nothing you added should require it.

File: lib/src/config_delayed_merge.cc [hunk 8]
```diff
@@ -0,0 +1,31 @@
+#include <internal/config_delayed_merge.hpp>
+#include <internal/config_delayed_merge_object.hpp>
+#include <internal/config_exception.hpp>
+
+using namespace std;
+
+namespace hocon {
+
+    config_delayed_merge::config_delayed_merge(shared_origin origin, std::vector<shared_value> stack) :
+        config_value(move(origin)), _stack(stack) {
```
Comment (MikaelSmith) [inline lib/src/config_delayed_merge.cc]: Should `move(stack)`.

File: lib/src/config_delayed_merge.cc [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+                throw config_exception("placed nested delayed_merge in a config_delayed_merge, should have consolidated stack");
+            }
+        }
+    }
+
+    config_value_type config_delayed_merge::value_type() const {
+        throw config_exception("called value_type() on value with unresolved substitutions, need to config#resolve() first, see API docs");
+    }
+
+    vector<shared_value> config_delayed_merge::unmerged_values() const {
```
Comment (MikaelSmith) [inline lib/src/config_delayed_merge.cc]: Yeah, this seems like it should return a `const &` to the vector, not a copy of the vector.

File: lib/src/objects/config_object.cc [hunk 10]
```diff
@@ -51,4 +52,42 @@ namespace hocon {
         return config_value_type::OBJECT;
     }
 
+    shared_ptr<const config_mergeable> config_object::with_fallback(shared_ptr<const config_mergeable> mergeable) const {
+        return dynamic_pointer_cast<const config_value>(config_value::with_fallback(mergeable));
```
Comment (MikaelSmith) [inline lib/src/objects/config_object.cc]: The dynamic_pointer_cast shouldn't be necessary. So we should be able to get rid of the override and fall back to the config_value version.

File: lib/src/objects/config_object.cc [hunk 11]
```diff
@@ -51,4 +52,42 @@ namespace hocon {
...
+            }
+
+            auto cv = dynamic_pointer_cast<const config_object>(v);
+            if (cv && cv->get_resolve_status() == resolve_status::RESOLVED && cv->is_empty()) {
+                // don't include empty files or the .empty()
+                // config in the description, since they are
+                // likely to be "implementation details"
+            } else {
+                origins.push_back(v->origin());
+                num_merged += 1;
```
Comment (MikaelSmith) [inline lib/src/objects/config_object.cc]: I know it mirrors the original, but origins.size() will always be equal to num_merged. We could do away with tracking it separately and check `origins.empty()` at line 84.

File: lib/src/simple_config_origin.cc [hunk 12]
```diff
@@ -54,6 +57,171 @@ namespace hocon {
         }
     }
 
+    shared_origin simple_config_origin::merge_origins(shared_origin a, shared_origin b) {
+        return merge_two(dynamic_pointer_cast<const simple_config_origin>(a), dynamic_pointer_cast<const simple_config_origin>(b));
+    }
+
+    shared_origin simple_config_origin::merge_origins(std::vector<shared_value> const& stack) {
+        vector<shared_origin> origins;
```
Comment (MikaelSmith) [inline lib/src/simple_config_origin.cc]: It's probably worth doing `origins.reserve(stack.size())`, to ensure we have the right amount of memory and don't have to realloc later.

File: lib/src/simple_config_origin.cc [hunk 13]
```diff
@@ -54,6 +57,171 @@ namespace hocon {
...
+        }
+
+        return merge_origins(origins);
+    }
+
+    shared_origin simple_config_origin::merge_origins(std::vector<shared_origin> const& stack) {
+        if (stack.empty()) {
+            throw config_exception("can't merge empty list of origins");
+        } else if (stack.size() == 1) {
+            return stack.front();
```
Comment (MikaelSmith) [inline lib/src/simple_config_origin.cc]: Is this equivalent to `stack.iterator().next()`?

File: lib/src/simple_config_origin.cc [hunk 14]
```diff
@@ -54,6 +57,171 @@ namespace hocon {
...
+        int merged_end_line;
+
+        origin_type merged_type;
+        if (a->_origin_type == b->_origin_type) {
+            merged_type = a->_origin_type;
+        } else {
+            merged_type = origin_type::GENERIC;
+        }
+
+        const string MERGE_OF_PREFIX { "merge of " };
```
Comment (MikaelSmith) [inline lib/src/simple_config_origin.cc]: Might as well make it `static`, to match the original.

File: lib/src/simple_config_origin.cc [hunk 15]
```diff
@@ -54,6 +57,171 @@ namespace hocon {
...
+
+        if (a_desc == b_desc) {
+            merged_desc = a_desc;
+
+            if (a->line_number() < 0) {
+                merged_start_line = b->line_number();
+            } else if (b->line_number() < 0) {
+                merged_start_line = a->line_number();
+            } else {
+                merged_start_line = max(a->_end_line_number, b->_end_line_number);
```
Comment (MikaelSmith) [inline lib/src/simple_config_origin.cc]: This should be `min`, not `max`.

File: lib/src/simple_config_origin.cc [hunk 16]
```diff
@@ -54,6 +57,171 @@ namespace hocon {
...
+
+            merged_end_line = max(a->_end_line_number, b->_end_line_number);
+        } else {
+            // this whole merge song-and-dance was intended to avoid this case
+            // whenever possible, but we've lost. Now we have to lose some
+            // structured information and cram into a string.
+
+            // description() method includes line numbers, so use it instead
+            // of description field.
+            string a_full = a->description();
```
Comment (MikaelSmith) [inline lib/src/simple_config_origin.cc]: This was already done for a_desc and b_desc above, I think we can just use them.

File: lib/src/simple_config_origin.cc [hunk 17]
```diff
@@ -54,6 +57,171 @@ namespace hocon {
...
+            merged_desc = MERGE_OF_PREFIX + a_full + "," + b_full;
+            merged_start_line = -1;
+            merged_end_line = -1;
+        }
+
+        string merged_resource;
+        if (a->_resource_or_null == b->_resource_or_null) {
+            merged_resource = a->_resource_or_null;
+        } else {
+            merged_resource = "";
```
Comment (MikaelSmith) [inline lib/src/simple_config_origin.cc]: Can remove the else clause, that's the default initialized behavior for merged_resource.

File: lib/inc/hocon/config_value.hpp [hunk 18]
```diff
@@ -163,8 +169,32 @@ namespace hocon {
...
+                                                 std::shared_ptr<const unmergeable> fallback) const;
+        shared_value merged_with_the_unmergeable(std::shared_ptr<const unmergeable> fallback) const;
+
+        shared_value merged_with_object(std::vector<shared_value> const& stack, shared_object fallback) const;
+        shared_value merged_with_object(shared_object fallback) const;
+
+        shared_value merged_with_non_object(std::vector<shared_value> const& stack, shared_value fallback) const;
+        shared_value merged_with_non_object(shared_value fallback) const;
+
+        shared_value construct_delayed_merge(shared_origin origin, std::vector<shared_value> stack) const;
```
Comment (MikaelSmith) [inline lib/inc/hocon/config_value.hpp]: The only reason this function exists is to be overridden by config_object. So it should be virtual.

File: lib/src/values/config_value.cc [hunk 19]
```diff
@@ -152,4 +154,112 @@ namespace hocon {
...
+                                                           std::shared_ptr<const unmergeable> fallback) const {
+        require_not_ignoring_fallbacks();
+
+        // if we turn out to be an object, and the fallback also does,
+        // then a merge may be required; delay until we resolve.
+        std::vector<shared_value> new_stack { stack };
+
+        auto unmerged_values = fallback->unmerged_values();
+        new_stack.insert(new_stack.end(), unmerged_values.begin(), unmerged_values.end());
+        return construct_delayed_merge(config_object::merge_origins(new_stack), new_stack);
```
Comment (MikaelSmith) [inline lib/src/values/config_value.cc]: Use `move(new_stack)`.

File: lib/src/values/config_value.cc [hunk 20]
```diff
@@ -152,4 +154,112 @@ namespace hocon {
...
+
+        if (get_resolve_status() == resolve_status::RESOLVED) {
+            // falling back to a non-object doesn't merge anything, and also
+            // prohibits merging any objects that we fall back to later.
+            // so we have to switch to ignoresFallbacks mode.
+            return with_fallbacks_ignored();
+        } else {
+            // if unresolved, we may have to look back to fallbacks as part of
+            // the resolution process, so always delay
+            return delay_merge(move(stack), move(fallback));
```
Comment (MikaelSmith) [inline lib/src/values/config_value.cc]: You don't want to `move` here, it's passing through a reference.

File: lib/inc/internal/simple_config_origin.hpp (hunk 7 above)
Comment (whopper) [inline lib/inc/internal/simple_config_origin.hpp]: Ah, looks like I added it because of one of the implementations of `simple_config_origin::merge_origins`, with:

```
    shared_origin simple_config_origin::merge_origins(std::vector<shared_value> const& stack) {
        vector<shared_origin> origins;
        for (auto& v : stack) {
            origins.push_back(v->origin());
        }
```
If I didn't include it, I'd get 'error: member access into incomplete type 'element_type' (aka 'const hocon::config_value') origins.push_back(v->origin());`

File: lib/src/values/config_value.cc [hunk 21]
```diff
@@ -152,4 +154,112 @@ namespace hocon {
...
+    }
+
+    shared_value config_value::merged_with_object(vector<shared_value> const& stack, shared_object fallback) const {
+        require_not_ignoring_fallbacks();
+
+        if (dynamic_pointer_cast<const config_object>(shared_from_this())) {
+            throw config_exception("Objects must reimplement merged_with_object");
+        }
+
+        return merged_with_non_object(move(stack), move(fallback));
```
Comment (MikaelSmith) [inline lib/src/values/config_value.cc]: Don't move, passing a reference.

File: lib/src/values/config_value.cc [hunk 22]
```diff
@@ -5,6 +5,8 @@
 #include <internal/objects/simple_config_object.hpp>
 #include <internal/simple_config_origin.hpp>
 #include <internal/container.hpp>
+#include <internal/config_delayed_merge.hpp>
+#include <internal/mergeable_value.hpp>
```
Comment (MikaelSmith) [inline lib/src/values/config_value.cc]: This header was removed, need to remove the `#include` (Travis is failing).

File: lib/inc/internal/simple_config_origin.hpp (hunk 7 above)
Comment (MikaelSmith) [inline lib/inc/internal/simple_config_origin.hpp]: Yes. Keep header files included in other header files to a minimum, otherwise you really kill compile time.

File: lib/src/values/config_value.cc (hunk 21 above)
Comment (whopper) [inline lib/src/values/config_value.cc]: Whoops, fixed!

File: lib/src/config_delayed_merge_object.cc [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    bool config_delayed_merge_object::is_empty() const {
+        // TODO
+        return {};
+    }
+
+    unordered_map<string, shared_value> const& config_delayed_merge_object::entry_set() const {
+        // TODO
+        return move(unordered_map<string, shared_value> {});
```
Comment (MikaelSmith) [inline lib/src/config_delayed_merge_object.cc]: You don't need move here, that's implicit for an object created in the return call.

**Comment by MikaelSmith** [comment]
Ok, I think that's good. :+1:
